src/function/scalar/string/regexp/regexp_extract_all.cpp

```diff
@@ -0,0 +1,217 @@
+			// the 'match_group' address is guaranteed to be bigger than that of the source
+			D_ASSERT((const char *)match_group.begin() >= string.GetDataUnsafe());
+			idx_t offset = match_group.begin() - string.GetDataUnsafe();
+			list_content[child_idx] = string_t(string.GetDataUnsafe() + offset, match_group.size());
+		}
+		current_list_size++;
+		if (startpos > input.size()) {
+			// Empty match found at the end of the string
+			break;
+		}
+	}
+	list_entry.length = current_list_size - list_entry.offset;
+	ListVector::SetListSize(result, current_list_size);
+}
+
+int32_t GetGroupIndex(DataChunk &args, idx_t row) {
+	if (args.ColumnCount() < 3) {
+		return 0;
+	}
+	UnifiedVectorFormat format;
+	args.data[2].ToUnifiedFormat(args.size(), format);
+	return ((int32_t *)format.data)[format.sel->get_index(row)];
+}
+
+void RegexpExtractAll::Execute(DataChunk &args, ExpressionState &state, Vector &result) {
+	auto &func_expr = (BoundFunctionExpression &)state.expr;
+	const auto &info = (RegexpBaseBindData &)*func_expr.bind_info;
+
+	auto &strings = args.data[0];
+	auto &patterns = args.data[1];
+	D_ASSERT(result.GetType().id() == LogicalTypeId::LIST);
+	auto &output_child = ListVector::GetEntry(result);
+
+	UnifiedVectorFormat strings_data;
+	strings.ToUnifiedFormat(args.size(), strings_data);
+
+	UnifiedVectorFormat pattern_data;
+	patterns.ToUnifiedFormat(args.size(), pattern_data);
+
+	ListVector::Reserve(result, STANDARD_VECTOR_SIZE);
+	// Reference the 'strings' StringBuffer, because we won't need to allocate new data
+	// for the result, all returned strings are substrings of the originals
+	output_child.SetAuxiliary(strings.GetAuxiliary());
+
+	// Avoid doing extra work if all the inputs are constant
+	idx_t tuple_count = args.AllConstant() ? 1 : args.size();
+	if (info.constant_pattern) {
+		auto &lstate = (RegexLocalState &)*ExecuteFunctionState::GetFunctionState(state);
+
+		auto group_count = lstate.constant_pattern.NumberOfCapturingGroups();
+		if (group_count == -1) {
+			throw InvalidInputException("Pattern failed to parse, error: '%s'", lstate.constant_pattern.error());
+		}
+
+		for (idx_t row = 0; row < tuple_count; row++) {
+			auto idx = strings_data.sel->get_index(row);
+			auto string = ((string_t *)strings_data.data)[idx];
+
+			auto &string_validity = strings_data.validity;
+			if (!string_validity.RowIsValid(idx)) {
+				// String is NULL, result is also NULL
+				auto result_data = FlatVector::GetData<list_entry_t>(result);
+				auto &result_validity = FlatVector::Validity(result);
+				result_data[row].length = 0;
+				result_data[row].offset = ListVector::GetListSize(result);
+				result_validity.SetInvalid(row);
+				continue;
+			}
+			int32_t group_index = GetGroupIndex(args, row);
+			// Get the groups
+			ExtractSingleTuple(string, lstate.constant_pattern, group_index, lstate.group_buffer, result, row);
+		}
+	} else {
+		RegexStringPieceArgs string_pieces;
+		for (idx_t row = 0; row < tuple_count; row++) {
```

Comment:
Can the code for the constant and non-constant pattern be unified? It looks like there is a lot of overlap.

For example, we could do something like this:
```cpp
unique_ptr<duckdb_re2::RE2> stored_re;
if (!info.constant_pattern) {
   // compile pattern string
}
auto &re = info.constant_pattern ? lstate.constant_pattern : *stored_re;
```